detection_rules/schemas/__init__.py

```diff
@@ -2,64 +2,141 @@
 # or more contributor license agreements. Licensed under the Elastic License
 # 2.0; you may not use this file except in compliance with the Elastic License
 # 2.0.
+import json
+import jsonschema
 
-from .base import TomlMetadata
 from .rta_schema import validate_rta_mapping
 from ..semver import Version
 from . import definitions
+from pathlib import Path
 
-# import all of the schema versions
-from .v7_8 import ApiSchema78
-from .v7_9 import ApiSchema79
-from .v7_10 import ApiSchema710
-from .v7_11 import ApiSchema711
-from .v7_12 import ApiSchema712
 
 __all__ = (
-    "all_schemas",
-    "available_versions",
+    "SCHEMA_DIR",
     "definitions",
     "downgrade",
-    "CurrentSchema",
     "validate_rta_mapping",
-    "TomlMetadata",
+    "all_versions",
 )
 
-all_schemas = [
-    ApiSchema78,
-    ApiSchema79,
-    ApiSchema710,
-    ApiSchema711,
-    ApiSchema712,
-]
-CurrentSchema = all_schemas[-1]
-available_versions = [cls.STACK_VERSION for cls in all_schemas]
+SCHEMA_DIR = Path(__file__).absolute().parent.parent.parent / "etc" / "api_schemas"
+migrations = {}
 
 
-def downgrade(api_contents: dict, target_version: str):
-    """Downgrade a rule to a target stack version."""
-    # truncate to (major, minor)
-    target_version_str = target_version
-    target_version = Version(target_version)[:2]
-    versions = set(Version(schema_cls.STACK_VERSION) for schema_cls in all_schemas)
-    role = api_contents.get("type")
+def all_versions():
+    """Get all known stack versions."""
+    return [str(v) for v in sorted(migrations)]
+
+
+def migrate(version):
+    """Decorator to set a migration."""
+    version = Version(version)
+
+    def wrapper(f):
+        assert version not in migrations
+        migrations[version] = f
+        return f
+
+    return wrapper
+
+
+@migrate("7.8")
+@migrate("7.9")
+@migrate("7.12")
```

Comment:
it feels error-prone to remember to add a decorator for each new version. 
 Can you build it dynamically for any schema version which does not have an explicit override.

Reply:
any thoughts on this?